src/index.ts

```diff
@@ -0,0 +1,137 @@
+import { requestsCount, requestDuration, requestStageDuration } from './metrics';
+import { parseUrlTemplate } from './utils';
+
+interface Headers {
+  [header: string]: number | string | string[] | undefined;
+}
+
+interface Options {
+  baseUrl: string;
+  headers: Headers;
+}
+
+interface Timings {
+  start?: number;
+  socket?: number;
+  lookup?: number;
+  connect?: number;
+  upload?: number;
+  response?: number;
+  end?: number;
+  error?: number;
+  phases: {
+    wait?: number;
+    dns?: number;
+    tcp?: number;
+    request?: number;
+    firstByte?: number;
+    download?: number;
+    total?: number;
+  };
+}
+
+interface Result {
+  timings?: Timings;
+  statusCode?: number;
+  body?: unknown;
+}
+
+type Method = 'GET' | 'PUT' | 'POST' | 'PATCH' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'TRACE' | 'CONNECT';
+
+export class HTTPClient {
+  baseUrl: string;
+  headers: Headers;
+
+  constructor(opts: Options) {
+    const { baseUrl, headers } = opts;
+    this.baseUrl = baseUrl;
+    this.headers = headers;
+  }
+
+  async request(
+    method: Method,
+    pathTemplate: string,
+    params = {},
+    body?: unknown,
+    options?: RequestInit
+  ): Promise<Result> {
+    const path = parseUrlTemplate(pathTemplate);
+    const url = new URL(path.expand(params), this.baseUrl);
+
+    const startTime = performance.now();
+
+    const defaultOptions: RequestInit = {
+      method: method.toUpperCase(),
+      headers: { ...this.headers, 'Content-Type': 'application/json' },
+      body: body ? JSON.stringify(body) : undefined,
+    };
+
+    const fetchOptions = { ...defaultOptions, ...options };
+
+    try {
+      const response = await fetch(url.toString(), fetchOptions);
+      const endTime = performance.now();
+
+      const result: Result = {
+        statusCode: response.status,
+        body: await response.json(),
+        timings: {
+          phases: {
+            total: endTime - startTime,
+          },
+        },
+      };
+
+      this._recordMetrics(result, method, pathTemplate);
+
+      return this._handlerResponse(result);
+    } catch (error) {
+      const endTime = performance.now();
+      const result: Result = {
+        statusCode: 0,
+        timings: {
+          phases: {
+            total: endTime - startTime,
+          },
+        },
+      };
+
+      this._recordMetrics(result, method, pathTemplate);
+
+      return this._handlerError(error as Error);
+    }
+  }
+
+  _recordMetrics(result: Result, method: string, pathTemplate: string) {
```

Comment:
_ is the javascript way to imply private.

typescript has private

```suggestion
  private recordMetrics(result: Result, method: string, pathTemplate: string) {
```

Reply:
fixed